src/bot.coffee

```diff
@@ -181,14 +177,14 @@ class SlackBot extends Adapter
       user = {} if !user
       user.room = channel.id
 
-      switch subtype
+      switch event.subtype
         when 'bot_message'
           @robot.logger.debug "Received message in channel: #{channel.name || channel.id}, from: #{user.name}"
 
           # prefer user over bot.
           # if both are set in the slack event, it represents an app or integration sending a message on behalf of a
           # user, so the user is the more appropriate value.
-          SlackTextMessage.makeSlackTextMessage(user || bot, undefined, undefined, event, channel, @robot.name, @client, (message) =>
+          SlackTextMessage.makeSlackTextMessage(user, undefined, undefined, event, channel, @robot.name, @client, (message) =>
```

Comment:
when this is a message with subtype `bot_message` (the case we are in), the user is only set when the app or integration is sending a message on behalf of the user (like when using `as_user: true`), otherwise it will be undefined, and in that case the `bot` should be the fallback, right?

Reply:
line 176 is setting the fallback to be the `bot` if the user is undefined, regardless of the `as_user` param

Reply:
maybe we need to back up a second and think about what that means.

the value of `bot` is jut a `bot_id` (`Bxxxx`), right? the value of `user` is the response of `users.info` (a full JSON object), right?

in either case, isn't doing the substitution on every message unnecessary? don't we only want to look into the bot object if we know the message is of subtype `bot_message`?